### PR TITLE
UCS/CALLBACKQ/TEST: Remove previous implementation of slowpath/oneshot

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Code cleanup of callbackq

### CODE REVIEW

**Comment by yosefe** [comment]
@rakhmets @brminich can you pls review?

File: test/gtest/ucs/test_callbackq.cc [hunk 1]
```diff
@@ -545,7 +414,7 @@ UCS_TEST_F(test_callbackq_noflags, oneshot_remove_another) {
     EXPECT_EQ(1, ctx1.count + ctx2.count);
 }
 
-UCS_TEST_F(test_callbackq_noflags, oneshot_remove_by_key) {
+UCS_TEST_F(test_callbackq, oeshot_remove_by_key) {
```
Comment (brminich) [inline test/gtest/ucs/test_callbackq.cc:417]: `oneshot`

File: src/ucs/datastruct/callbackq.c [hunk 2]
```diff
@@ -41,354 +49,342 @@ typedef struct ucs_callbackq_oneshot_elem {
...
-}
+
+static void ucs_callbackq_proxy_enable(ucs_callbackq_t *cbq);
+static void ucs_callbackq_proxy_disable(ucs_callbackq_t *cbq);
 
 static void ucs_callbackq_enter(ucs_callbackq_t *cbq)
 {
-    ucs_recursive_spin_lock(&ucs_callbackq_priv(cbq)->lock);
+    ucs_recursive_spin_lock(&cbq->priv->lock);
+    ucs_log_indent_level(UCS_LOG_LEVEL_TRACE_FUNC, +1);
```
Comment (rakhmets) [inline src/ucs/datastruct/callbackq.c:91]: As I see we use just `1` in case of the indentation increasing, so `+` can be removed here.

**Comment by yosefe** [comment]
@brminich @rakhmets fixed

File: src/ucs/datastruct/callbackq.c [hunk 3]
```diff
@@ -41,324 +49,243 @@ typedef struct ucs_callbackq_oneshot_elem {
...
+       in this location */
+    ucs_callbackq_fast_elem_set(cbq, idx, NULL, cbq, UCS_CALLBACKQ_ID_NULL);
 }
 
 /*
  * @param [in]  id  ID to release in the lookup array.
  * @return index which this ID used to hold.
  */
-int ucs_callbackq_put_id(ucs_callbackq_t *cbq, int id)
+unsigned ucs_callbackq_put_id(ucs_callbackq_t *cbq, int id)
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucs/datastruct/callbackq.c:122]: since it's internal function, cbq is not documented to not state the obvious..

**Review by brminich — APPROVED** [review]

File: src/uct/base/uct_worker.c [hunk 4]
```diff
@@ -107,29 +120,87 @@ void uct_worker_progress_remove_all(uct_priv_worker_t *worker,
...
+    if (flags & UCS_CALLBACKQ_FLAG_ONESHOT) {
+        /*
+         * Implement UCS_CALLBACKQ_FLAG_ONESHOT flag, for backward compatibility
+         * after removing flags parameter from ucs_callbackq. Since we have to
+         * return a callback id, we cannot use ucs_callbackq_add_oneshot() API,
+         * and we emulate oneshot behavior by making the callback remove itself.
+         */
+        ctx = ucs_malloc(sizeof(*ctx), "uct_worker_oneshot_cb_ctx");
+        if (ctx == NULL) {
+            goto out;
```
Comment (rakhmets) [inline src/uct/base/uct_worker.c:157]: Please add error logging when memory allocation fails.

File: src/uct/base/uct_worker.c [hunk 5]
```diff
@@ -107,29 +120,87 @@ void uct_worker_progress_remove_all(uct_priv_worker_t *worker,
...
+        ctx->func = func;
+        ctx->arg  = arg;
+        ctx->cbq  = &worker->super.progress_q;
+        ctx->id   = ucs_callbackq_add_safe(&worker->super.progress_q,
+                                           uct_worker_oneshot_callback_proxy,
+                                           ctx);
+
+        id = ctx->id + UCT_WORKER_ONESHOT_ID_START;
+        ucs_assertv(id >= UCT_WORKER_ONESHOT_ID_START, "id=%d", id);
+
```
Comment (rakhmets) [inline src/uct/base/uct_worker.c:169]: Please remove extra empty line.

File: test/gtest/uct/test_progress.cc [hunk 6]
```diff
@@ -17,6 +17,23 @@ class test_uct_progress : public uct_test {
         uct_test::init();
         m_entities.push_back(create_entity(0));
     }
+
+protected:
+    uct_worker_h worker(unsigned index = 0) {
```
Comment (rakhmets) [inline test/gtest/uct/test_progress.cc:22]: Please review code format issues in test_callbackq.cc and in test_progress.cc. There are several issues with brace location.

File: test/gtest/uct/test_progress.cc [hunk 7]
```diff
@@ -30,14 +47,46 @@ UCS_TEST_P(test_uct_progress, random_enable_disable) {
...
+                                      UCS_CALLBACKQ_FLAG_ONESHOT, &prog_id);
+    EXPECT_NE(UCS_CALLBACKQ_ID_NULL, prog_id);
+
+    uct_worker_progress_unregister_safe(worker(), &prog_id);
+    EXPECT_EQ(UCS_CALLBACKQ_ID_NULL, prog_id);
+
+    unsigned count = progress();
+    EXPECT_EQ(0, count);
+    EXPECT_EQ(0, m_count);
+
```
Comment (rakhmets) [inline test/gtest/uct/test_progress.cc:88]: Please remove extra empty line.

File: test/gtest/uct/test_progress.cc [hunk 8]
```diff
@@ -30,14 +47,46 @@ UCS_TEST_P(test_uct_progress, random_enable_disable) {
...
+                                      UCS_CALLBACKQ_FLAG_ONESHOT, &prog_id);
+    EXPECT_NE(UCS_CALLBACKQ_ID_NULL, prog_id);
+
+    EXPECT_EQ(0, m_count);
+    unsigned count1 = progress();
+    EXPECT_GE(count1, 1);
+
+    /* The callback should be removed by now */
+    unsigned count2 = progress();
+    EXPECT_EQ(0, count2);
```
Comment (rakhmets) [inline test/gtest/uct/test_progress.cc:71]:
```suggestion
    unsigned count = progress();
    EXPECT_GE(count, 1);

    /* The callback should be removed by now */
    count = progress();
    EXPECT_EQ(0, count);
```

**Comment by yosefe** [comment]
@rakhmets fixed

**Review by rakhmets — APPROVED** [review]

**Review by rakhmets — APPROVED** [review]

**Pull request merged**